internal/chain/engine.go

```diff
@@ -402,6 +427,78 @@ func IsCommitteeSelectionBlock(chain engine.ChainReader, header *block.Header) b
 	return isBeaconChain && header.IsLastBlockInEpoch() && inPreStakingEra
 }
 
+// shouldPruneStaleStakingData checks (1) we are in the beacon chain, and
+// (2) ChainConfig returns true for IsNoNilDelegations
+func shouldPruneStaleStakingData(
+	chain engine.ChainReader,
+	header *block.Header,
+) bool {
+	return header.ShardID() == shard.BeaconChainShardID &&
+		header.IsLastBlockInEpoch() &&
+		chain.Config().IsNoNilDelegations(header.Epoch())
```

Comment:
Note the pruning only happens at the end of the IsNoNilDelegations epoch. So the new reward calculation logic only need to happen in the blocks with IsNoNilDelegations epoch + 1

Reply:
Thanks, fixed.